keps/sig-apps/20190318-Graduate-CronJob-to-Stable.md

```diff
@@ -0,0 +1,98 @@
+---
+title: graduate-cronjob-to-stable
+authors:
+  - "@barney-s"
+owning-sig: sig-apps
+participating-sigs:
+  - sig-scheduling
+reviewers:
+  - "@liggitt"
+  - "@kow3ns"
+  - "@janetkuo"
+  - “@mortent”
+approvers:
+  - "@kow3ns"
+  - "@liggitt"
+editor: TBD
+creation-date: 2019-04-18
+last-updated: 2019-04-18
+status: implementable
+see-also:
+  - 
+replaces:
+superseded-by:
+---
+
+# Graduate CronJob to stable
+
+## Table of Contents
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Implementation Notes](#implementation-notes)
+  - [Constraints](#constraints)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+
+## Summary
+
+[CronJob](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/) is a Kubernetes API that creates Job object on a schedule specified by a cron spec. It is in beta status for a long time. This document lays out the plan to promote it to stable.
+
+## Motivation
+
+CronJob definition has been stable for the last few releases and is useful to run periodic tasks in kubernetes cluster. This API adds the ability to add cron facility to a cluster. We feel the API is ready to be promoted to Stable and be supported longterm by the community.
+
+### Goals
+
+* Plan to promote CronJob API to stable version.
+
+### Non-Goals
+
+* Changing API field or meaning
+
+## Proposal
+
+### Implementation Notes
+
+In the current implementation, the controller: 
+
+1. syncs all CronJob objects [every 10 seconds](https://github.com/kubernetes/kubernetes/blob/30165e40ddfbe75fddc575c14294c6b540361078/pkg/controller/cronjob/controller.go#L98). 
+2. Using pager library, gets all Pods and all CronJobs and [processes them one by one](https://github.com/kubernetes/kubernetes/blob/30165e40ddfbe75fddc575c14294c6b540361078/pkg/controller/cronjob/controller.go#L144)
+
+The current implementation can be improved to reduce the need to list all Pod and CronJob objects frequently to reconcile. Instead we should replace it with an Informers and WorkQueue based architecture. Preferable we should be sharing the same informer cache as the Job controller uses.
+
+This is required to:
+
+1. Reduce the potential scale issues when using lots of pods and CronJob objects (not measured).  
+2. Reduce load on API server in such cases.
+3. Reduce memory usage when listing all Pods and CronJobs in every sync loop.
```

Comment:
the scaling issues with cronjobs are currently quite severe. I'd like to see scale targets (both in terms of cronjob count per namespace and cluster-wide, and in terms of frequency of launched jobs), and a test plan to exercise those targets included in the KEP.

Reply:
If we do this sig scalability should be a reviewer, right?

Reply:
yes, that would be recommended. see discussion in https://github.com/kubernetes/enhancements/pull/1015 for prior art for framing scale targets for a specific resource

Reply:
added a scaling section.